### DISCUSSION TITLE
Adding display/color modes to PineTime

### DISCUSSION THREAD

**Discussion by wilsonjholmes** [InfiniTimeOrg/InfiniTime#1612] [Ideas]
## Brief overview on how this came about: 
 
I initially submitted PR [#633](https://github.com/JF002/InfiniTime/pull/633) as I wanted to make the watch more astronomer-friendly by adding a way for the flashlight to be changed to red as a way to help preserve night vision when doing astrophotography and the like. That PR ended up raising the question of where individual app settings should be accessed on the watch: In a settings "pane", or through the app itself. That is kinda its [own discussion now](https://github.com/JF002/InfiniTime/pull/636#issuecomment-912052329), but as [tmilburn](https://github.com/tmilburn) and [geekbozu](https://github.com/geekbozu) brought up in the discussion on that PR, It might be better for the whole UI of the watch to be changed to a sort of "redscale" so that navigating the watch could also be done without messing up time acquired night vision/sensitivity.

## Where things are now

GeekBoy1011 on the discord helped me get the display buffer bit-mathed in DrawDisplay before it was actually drawn, and that was very promising, still very rough, but here is a proof of concept.
![main](https://user-images.githubusercontent.com/59266497/132147941-e7d22d60-d974-4f2c-8ad6-cdc6cb8fee1f.jpeg)
![quicksett](https://user-images.githubusercontent.com/59266497/132147940-3c91f875-bcba-4b3a-84c7-232bfd594db3.jpeg)
![apps](https://user-images.githubusercontent.com/59266497/132147939-d53f4d53-02e2-44d6-aacf-d7265c76ed78.jpeg)
![accel](https://user-images.githubusercontent.com/59266497/132147937-268c49f4-7f9c-4fd1-bb04-603f943cd808.jpeg)
Currently only red colors showing, and any color that is not red is "converted" so you are still able to see things that are pure green or blue in the UI. This is in progress now and looking nice, but there are some issues thus far to think about.

## In need of ideas on the best way to move forward with this

This change has the possibility to wreck contrast in parts of the UI if certain colors are used in elements that are stacked on top of one another. A pure red will display the same as a "pure" purple. I am not sure what the best way to mitigate this is:
- While it may take more work I was wondering if it would just be better to change all of the UI to use defined color values, and those colors could be swapped out by a setting depending on what display mode you are in (e.g. Normal, Night-Mode, Greyscale)... Kinda like we already have `settingsController.GetPTSColor*()` in `PineTimeStyle.cpp`. We could, by this same method have a way for a completely custom set of colors to be defined for a redscale UI that would still have proper contrast...
- I have not really looked much into changing the gamma of the display, but I know that is another option to look into.
- We could just use the same bit masking/color conversion thing as show in the pictures and make sure that all colors in UI have significant enough contrast.
- I'm sure there are some things I have not though of yet... eager to hear input and maybe some other ideas.

**Comment by Riksu9000** [comment]
I think the UI should have high contrast for usability and accessibility anyway, so just a filter should be good. Here's a color temperature conversion algorithm that can be used for "low blue light modes" https://tannerhelland.com/2012/09/18/convert-temperature-rgb-algorithm-code.html

**Comment by evergreen22** [comment]
I like the idea of a 'night mode' in red that does not mess up night vision. I can think of other places this would be useful such as theatre, film, old school photography, and of course ninja.

The color settings are sprinkled all through the code now and it would be nice to have a common interface to set the screen colors. Perhaps a collection of themes that each 'screen' could index into to get the color setup. Then a color picker could just modify a 'user theme' for personalization.

**Comment by wilsonjholmes** [comment]
> I like the idea of a 'night mode' in red that does not mess up night vision. I can think of other places this would be useful such as theatre, film, old school photography, and of course ninja.
> 
> The color settings are sprinkled all through the code now and it would be nice to have a common interface to set the screen colors. Perhaps a collection of themes that each 'screen' could index into to get the color setup. Then a color picker could just modify a 'user theme' for personalization.

I was thinking that would be slick, but I am new to the project, and would need some help understanding the different types of color categories to be used in a theme to cover all parts of the UI...

**Reply by NitroNils** [reply]
I was guided here as I'm curious about a "redshift" for the display.
Has there been any progress? Uniting color adjustments for easier customization like night vision conservation or other needs...

**Reply by sammyj85** [reply]
A redshift for night time would be amazing. I can customise the watch face colours to orange, which is a good approximation, but there is a big shock seeing the pure white in the menu, or worse, with the flashlight.

**Comment by hubmartin** [comment]
With PWM backlight PR you can change backlight in range fron 0 - 10000, from darkness to full brightnness.
https://github.com/JF002/InfiniTime/pull/575

**Comment by joelghill** [comment]
I've been thinking lately that it would be really cool to upload user themes to the watch and then select them from a list.

We would need to predefine a standard set of colours for backgrounds, text, accents, etc, in a single place and update everything to reference those values.

Is there such a thing as a standard common colour theme protocol that could be adopted? It would be cool to support existing themes, like from vim plugins or something.

**Reply by owenfromcanada** [reply]
There's no standard that I'm aware of--but we can crib from other similar projects and start working out a practical list of colors to map.

For example, in Android, there are a few values that the underlying system uses, including default fore/background colors, primary and secondary accent colors, and some UI colors (for example, default button colors, including "enabled" and "disabled" colors).

Would it be helpful if I attempted to put together a proposed list of colors? I'd be happy to work on this, but it'll end up being pretty invasive (as all the other apps would eventually need to be updated).
